Added support for stdfunc='mad_std' in sigma clipping

astrofrog opened the pull request:
This includes the required changes to make sure that the new C implementation can recognize this, leading to significant speedup in some cases:

![Screenshot_2021-04-29_15-36-01](https://user-images.githubusercontent.com/314716/116569266-23771400-a901-11eb-91a0-5c0b983817e8.png)

cc @keflavich FYI

pllim commented:
Any chance you can provide a snippet that people can copy-paste if they want to profile this on their own?

astrofrog commented:
Yes here is a notebook: https://gist.github.com/astrofrog/002ffc3562f4fb647963207c0b488471

astrofrog commented:
The 32-bit failure is real, will try and investigate later but feel free to review this anyway in the mean time!

mhvk commented on astropy/stats/sigma_clipping.py:
```diff
@@ -154,7 +160,7 @@ class SigmaClip:
         .. _bottleneck:  https://github.com/pydata/bottleneck
 
-    stdfunc : {'std'} or callable, optional
+    stdfunc : {'std', 'mad_std'} or callable, optional
```
Slightly unrelated, but the text below mentions `bottleneck`, but not the built-in fast method.

astrofrog replied:
I've tried to clarify this

mhvk commented on astropy/stats/sigma_clipping.py:
```diff
@@ -590,7 +600,7 @@ def __call__(self, data, axis=None, masked=True, return_bounds=False,
 
         # Shortcut for common cases where a fast C implementation can be used.
-        if self.cenfunc in ('mean', 'median') and self.stdfunc == 'std' and axis is not None and not self.grow:
+        if self.cenfunc in ('mean', 'median') and self.stdfunc in ('std', 'mad_std') and axis is not None and not self.grow:
```
break line?

larrybradley replied:
Yes, please.  The flake8 test uses a maximum line length of 100.  This line is longer than that.

astrofrog replied:
Done

mhvk commented on astropy/stats/src/compute_bounds.c:
```diff
@@ -1,41 +1,58 @@
 
-      *lower_bound = median - sigma_lower * std;
-      *upper_bound = median + sigma_upper * std;
+      buffer2 = malloc(count * sizeof(double*));
```
No need to allocate the buffer again!?

astrofrog replied:
Oops yes leftover code

astrofrog replied:
Fixed

mhvk commented on astropy/stats/src/compute_bounds.c:
```diff
@@ -1,41 +1,58 @@
+                                  int use_mad_std, int maxiters, double sigma_lower,
                                   double sigma_upper, double *lower_bound,
-                                  double *upper_bound) {
+                                  double *upper_bound, double buffer2[]) {
```
I wonder if for clarity it would be better to call the first `buffer` something else, like `data[]` or `data_buffer[]`, and this one just `buffer` or `mad_buffer`

astrofrog replied:
Done

mhvk commented on astropy/stats/src/fast_sigma_clip.c:
```diff
@@ -100,17 +103,30 @@ static void _sigma_clip_fast(
         return;
     }
+
+    // buffer2 is available to be used by compute_sigma_clipped_bounds
```
Ideally, we only allocate this when actually needed. The data arrays can be quite large. I'd suggest setting it to NULL and then inside the loop check whether `use_mad` is set, and if so replace a NULL by a memory allocation. Obviously, in that case, the freeing also needs to check whether the buffer was allocated in the first place.

astrofrog replied:
I am running out of time for the feature freeze so can we get this in and the follow up can always go in later as it would not be an additional feature just an optimization of existing code?

mhvk commented on astropy/stats/tests/test_sigma_clipping.py:
```diff
@@ -463,3 +464,50 @@ def test_sigma_clip_dtypes(dtype):
+                                maxiters=1, sigma=5, masked=False, axis=0)
+    assert_equal(result_mad_std, [1, np.nan, 4, 3, np.nan])
+
+    # And now test with a larger array and compare with Python mad_std function
```
Maybe make this a separate test.

astrofrog replied:
Done

mhvk commented on docs/changes/stats/11664.feature.rst:
```diff
@@ -0,0 +1,3 @@
+Added the ability to specify stdfunc='mad_std' when doing sigma clipping,
+which leads to significant performance improvements if cenfunc is 'mean'
```
maybe "which will use a built-in function and give significant ..."

astrofrog replied:
Fixed

mhvk reviewed (commented):
Overall, this looks good. Some comments on the buffer in the implementation, plus a few nitpicks.

astrofrog commented:
@mhvk - I have addressed all your comments except about avoiding the memory allocation if possible. Can I defer that to a follow-up PR next week to make sure the new feature gets in by tonight?

@larrybradley - feel free to review!

astrofrog commented:
The 32-bit failure was due to the extra memory allocation - should be fixed now

mhvk commented on astropy/stats/sigma_clipping.py:
```diff
@@ -144,21 +150,14 @@ class SigmaClip:
-
-        .. _bottleneck:  https://github.com/pydata/bottleneck
+        center value for the clipping. If using a callable function/object and
+        the ``axis`` keyword is used, then it must be callable that can ignore
```
"then it must be callable that can ignore" -> "then it must be able to ignore"

astrofrog replied:
Fixed

mhvk commented on astropy/stats/sigma_clipping.py:
```diff
@@ -144,21 +150,14 @@ class SigmaClip:
-        result in the best performance.  If using a callable
+        standard deviation about the center value. If using a callable
         function/object and the ``axis`` keyword is used, then it must
         be callable that can ignore NaNs (e.g., `numpy.nanstd`) and has
```
And same (so this existed before). If you wish you can also just insert "a" before "callable", but I think my suggestion above is better.

astrofrog replied:
Fixed

mhvk commented on astropy/stats/sigma_clipping.py:
```diff
@@ -673,21 +688,14 @@ def sigma_clip(data, sigma=3, sigma_lower=None, sigma_upper=None, maxiters=5,
-
-        .. _bottleneck:  https://github.com/pydata/bottleneck
+        center value for the clipping. If using a callable function/object and
+        the ``axis`` keyword is used, then it must be callable that can ignore
```
same again

astrofrog replied:
Fixed

mhvk commented on astropy/stats/sigma_clipping.py:
```diff
@@ -673,21 +688,14 @@ def sigma_clip(data, sigma=3, sigma_lower=None, sigma_upper=None, maxiters=5,
-        result in the best performance.  If using a callable
+        standard deviation about the center value. If using a callable
         function/object and the ``axis`` keyword is used, then it must
         be callable that can ignore NaNs (e.g., `numpy.nanstd`) and has
```
And again.

mhvk commented on astropy/stats/sigma_clipping.py:
```diff
@@ -843,21 +861,14 @@ def sigma_clipped_stats(data, mask=None, mask_value=None, sigma=3.0,
-        and has an ``axis`` keyword to return an array with axis
-        dimension(s) removed.  The default is ``'median'``.
+        center value for the clipping. If using a callable function/object and
+        the ``axis`` keyword is used, then it must be callable that can ignore
```
And again

astrofrog replied:
Fixed

mhvk commented on astropy/stats/sigma_clipping.py:
```diff
@@ -843,21 +861,14 @@ def sigma_clipped_stats(data, mask=None, mask_value=None, sigma=3.0,
-        result in the best performance.  If using a callable
+        standard deviation about the center value. If using a callable
         function/object and the ``axis`` keyword is used, then it must
         be callable that can ignore NaNs (e.g., `numpy.nanstd`) and has
```
idem

astrofrog replied:
Fixed

mhvk commented on astropy/stats/src/compute_bounds.h:
```diff
@@ -1,8 +1,8 @@
 #define __COMPUTE_BOUNDS_H__
 
-void compute_sigma_clipped_bounds(double buffer[], int count, int use_median,
+void compute_sigma_clipped_bounds(double buffer[], int count, int use_median, int use_mad_std,
```
Update the names of `buffer` and `buffer2` here too

astrofrog replied:
Fixed

mhvk commented on astropy/stats/src/fast_sigma_clip.c:
```diff
@@ -99,39 +102,54 @@ static void _sigma_clip_fast(
         return;
     }
+
+    // mad_buffer is available to be used by compute_sigma_clipped_bounds
```
So, I'd move this inside the loop, as follows:
```
if (((npy_bool *)use_mad_std) && mad_buffer == NULL) {
...
}
```

astrofrog replied:
Done

mhvk commented on astropy/stats/src/fast_sigma_clip.c:
```diff
@@ -99,39 +102,54 @@ static void _sigma_clip_fast(
     }
-    PyArray_free((void *)buffer);
+    PyArray_free((void *)data_buffer);
+    PyArray_free((void *)mad_buffer);
```
And here
```
if (mad_buffer != NULL) {
    PyArray_free(...)
}
```

astrofrog replied:
Done

mhvk reviewed (commented):
@astrofrog - looks good, but one real suggestion still to move the memory allocation for `mad` inside the loop, so it gets used only when actually needed (since the call is slow, the extra time in the if-statement is negligible).

Also, not really you, but might as well fix something in the docstring... (see inline)

mhvk reviewed (approved):
All great now!
